### PR TITLE
VideoPress: Introduce `VideoRow` component

### LINKED ISSUE #25742: VideoPress components: VideoRow
Image

Image

### PR DESCRIPTION
**Pull request by renatoagds**
This PR introduce the `VideoRow` component, that will be used to list the videos library.

Closes #25742 

#### Changes proposed in this Pull Request:

* Introduce new component `VideoRow`.
* Support `ref` at `Text` component.

#### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

N/A

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

* Start `Storybook`
* Enable `QUICK_ACTIONS` - https://github.com/Automattic/jetpack/blob/1623493a658baf47c2d3938cb2e90fe491295230/projects/packages/videopress/src/client/admin/components/video-row/index.tsx#L67
* Navigate to `VideoPress/Video Row` story.
* Interact with `mobile` and `desktop` version.

### DEMOS

https://user-images.githubusercontent.com/1663717/186263220-d0abe4e5-cb65-4e42-9e19-e69f27f21b4d.mov

https://user-images.githubusercontent.com/1663717/186263252-2814c9c3-20bd-47b7-b85f-4b5abedd3655.mov

https://user-images.githubusercontent.com/1663717/186263262-426a0236-a6cd-4d05-81b6-126f6a232d69.mov

### CODE REVIEW

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+		setShowActions( true );
+	};
+
+	const handleLeave = () => {
+		setShowActions( false );
+	};
+
+	const handleClick = () => {
+		toggleExpand();
+	};
```
Comment (retrofox) [inline projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: It looks confusing.

`handleClick` is called once, via the `infoWrapperActions` instance.
it calls to `toggleExpand();`, which calls the `setExpanded()`. 
Is it possible to simplify it?

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+			className={ classNames( styles[ 'video-row' ], {
+				[ styles.pressed ]: keyPressed,
+				[ styles.small ]: isSmall,
+			} ) }
+			role="button"
+			tabIndex={ 0 }
+		>
+			<div
+				className={ classNames( styles[ 'info-wrapper' ], { [ styles.small ]: isSmall } ) }
+				{ ...infoWrapperActions }
```
Comment (retrofox) [inline projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: why not simply set the onClick property? It results in a less clear understanding of the code. Maybe it should replaced by something like the following:
```suggestion
				onClick= { isSmall ? handleClick : null }
```

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+		onMouseLeave: isSmall ? null : handleLeave,
+		'aria-label': wrapperAriaLabel,
+	};
+
+	// If small screen, expand on click at info wrapper
+	const infoWrapperActions = { onClick: isSmall ? handleClick : null };
+
+	return (
+		<div
+			{ ...wrapperActions }
```
Comment (retrofox) [inline projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: I don't see the need to create an object and then destructuring it.

**Comment by retrofox** [comment]
Design-wise, I see the following inconsistencies:

design | implementation | .
------|------|-----
Screen Shot 2022-08-24 at 09 58 37|Screen Shot 2022-08-24 at 09 59 34| wrong height
image | image | wrong padding
image|Screen Shot 2022-08-24 at 10 08 45|trash icon distance

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+						<span>{ privacyLabel }</span>
+						<span>
+							{ isPrivate
+								? __( 'Private', 'jetpack-videopress-pkg' )
+								: __( 'Public', 'jetpack-videopress-pkg' ) }
+						</span>
+					</>
+				) }
+			</div>
+			<div className={ styles.duration }>
```
Comment (retrofox) [inline projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: `duration` CSS class is not defined.

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+				<span>{ isSmall && durationLabel }</span>
+				<span>{ duration }</span>
+			</div>
+			{ Number.isFinite( plays ) && (
+				<div className={ styles.plays }>
```
Comment (retrofox) [inline projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: same, `plays` CSS class is not defined

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 4 above)
Comment (retrofox) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: We should replace this div with a Text:

```suggestion
			<Text variant="body" className={ styles.duration }>
```

btw font-size is not correct: `16px` vs `13px`

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+import classNames from 'classnames';
+import { useState, useRef } from 'react';
+import privacy from './privacy-icon';
+import styles from './style.module.scss';
+
+const millisecondsToMinutesAndSeconds = ( milliseconds: number ) => {
+	const minutes = Math.floor( milliseconds / 60000 );
+	const seconds = Math.floor( ( milliseconds % 60000 ) / 1000 );
+	return `${ minutes }:${ seconds < 10 ? '0' : '' }${ seconds }`;
+};
```
Comment (dhasilva) [inline projects/packages/videopress/src/client/admin/components/video-row/index.tsx:15]: This seems like it should be an imported helper function.

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+			<Icon icon={ icon } />
+			{ showPopover && <PopoverWithAnchor anchorRef={ anchorRef }>{ children }</PopoverWithAnchor> }
+		</div>
+	);
+};
+
+const QuickActions = ( { button }: { button: React.ReactNode } ) => {
+	// Hiding it based on Design request:
+	// https://github.com/Automattic/jetpack/issues/25742#issuecomment-1223123815
+	const HIDE_QUICK_ACTIONS = false;
```
Comment (dhasilva) [inline projects/packages/videopress/src/client/admin/components/video-row/index.tsx]:
```suggestion
	const HIDE_QUICK_ACTIONS = true;
```

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 6 above)
Comment (retrofox) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx:15]: we have a few helpers defined in the Jetpack plugin if I'm not wrong. We can migrate them and update the code, but surely IMO in a follow-up one.

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+	onClickEdit?: () => void;
+} ) => {
+	const [ isSmall ] = useBreakpointMatch( 'sm' );
+	const [ showActions, setShowActions ] = useState( false );
+	const [ keyPressed, setKeyDown ] = useState( false );
+	const [ expanded, setExpanded ] = useState( false );
+	const textRef = useRef( null );
+
+	const durationInMinutesAndSeconds = millisecondsToMinutesAndSeconds( duration );
+	const uploadDateFormatted = dateI18n( 'M j, Y', uploadDate, null );
```
Comment (dhasilva) [inline projects/packages/videopress/src/client/admin/components/video-row/index.tsx:182]: Question: Is this format `M j, Y` applied in all languages?

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+	const isSpaceOrEnter = code => code === 'Space' || code === 'Enter';
+	const isEllipsisActive = textRef?.current?.offsetWidth < textRef?.current?.scrollWidth;
+	const showTitleLabel = ! isSmall && isEllipsisActive;
```
Comment (dhasilva) [inline projects/packages/videopress/src/client/admin/components/video-row/index.tsx:184]: There seems to be no way to see the full title on mobile when the text is cropped. Maybe only `isEllipsisActive` should be enough?

**Comment by renatoagds** [comment]
@retrofox about the design part, there’s inconsistencies across figma, I’ll reach out design and we can improve (if needed) later. Let’s follow this way for now.

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 9 above)
Comment (renatoagds) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx:184]: Over doesn’t looks well at mobile, that’s the reason I deactivated it, we should reach out to design for this one.

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 6 above)
Comment (renatoagds) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx:15]: Yes, let’s move at some moment 👍

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 2 above)
Comment (renatoagds) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: Don’t think like that, but I will update it

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 4 above)
Comment (renatoagds) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: Forgot to remove them, working on it

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 1 above)
Comment (renatoagds) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: Yes

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 8 above)
Comment (retrofox) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx:155]: > @retrofox about the design part, there’s inconsistencies across figma, I’ll reach out design and we can improve (if needed) later. Let’s follow this way for now.

Where there are the inconsistencies? The row height is 90px in all instances that I took a look at, for instance.
Why we are going to implement something that is not consistent with the design? Even more, we can clarify the inconsistencies just here, asking designer folks. This is the proper moment to do it.

**Comment by retrofox** [comment]
> @retrofox about the design part, there’s inconsistencies across figma, I’ll reach out design and we can improve (if needed) later. Let’s follow this way for now.

Where there are the inconsistencies? The row height is 90px in all instances that I took a look at, for instance.
Why we are going to implement something that is not consistent with the design? Even more, we can clarify the inconsistencies just here, asking designer folks. This is the proper moment to do it.

**Comment by renatoagds** [comment]
@retrofox let's jump on it:

1) *Wrong height*

If you take a look at all examples of poster images (a.k.a thumbnail), it doesn't follow the same height for all of them; Image below:

image

So, if we took the minimum one (_59.41px_) and round it for _60px_ (what I did), and include the top and bottom padding (_24px_), we have: `60 + 24 + 24 = 108`, that's exactly the height implemented right now.

To implement the height at `90px` I need to reduce the `height` of the poster to `42px`, which generates a new inconsistency.

Probably is related with the white space at the posters examples.

2) *Wrong padding*

This one is related only to `left` padding, and you're right, but the correct padding is `16px`, anyway, is missing the checkbox. I'll fix it.

3) *Trash icon distance*

The distance at *Figma* is `30px`, which is not an `8px` multiple, if you do the math `(30 / 8)` is 3.75, so I rounded it up.

image

Anyway, I will round it down to make it close to Figma.

> Even more, we can clarify the inconsistencies just here, asking designer folks. This is the proper moment to do it.

Sure, let's do it. I just wanted to move forward with this one, as we do normally, and jump to talk with them later, but if that's the way you want to follow, np.

**Comment by retrofox** [comment]
> Sure, let's do it. I just wanted to move forward with this one, as we do normally, and jump to talk with them later, but if that's the way you want to follow, np.

👍 ok

**Comment by keoshi** [comment]
> Wrong height

Just a note that the screenshot you posted above is indicating the _image size inside_ each thumbnail block. The block is actually acting as a frame that clips those images, and all thumbnails should follow a size of `90 x 50 px`:

image

> Wrong padding

If no checkbox is visible, the left padding for the row should be 8px.

image

> Trash icon distance

The trash icon (like all others except for the eye) is coming directly from the Core WordPress library. Not sure if we can reuse that or are still adding icon manually, but wanted to note that.

**Comment by renatoagds** [comment]
@keoshi thanks, just a few points:

> Just a note that the screenshot you posted above is indicating the image size inside each thumbnail block. The block is actually acting as a frame that clips those images, and all thumbnails should follow a size of 90 x 50 px:

With that, we'll have a height of `98px`, that's different from `90px` reported from @retrofox, just to confirm that's ok to follow with it.

It's related too with the case that the `padding` applied is `20px` at Figma, which doesn't follow our `8px` rule

> The trash icon (like all others except for the eye) is coming directly from the Core WordPress library. Not sure if we can reuse that or are still adding icon manually, but wanted to note that.

We're using from `core`, but it doesn't implement the `margin-right`, it's made by us, based on that, should I follow the `8px` rule as normal? Which will move it to `32px (8 x 4)` instead of `30px`. Is that ok?

**Comment by retrofox** [comment]
> With that, we'll have a height of `98px`, that's different from `90px` reported from @retrofox, just to confirm that's ok to follow with it.

sorry, it's `50px` (frame height) + `20px` + `20px` (paddings) = `90px`.

**Comment by renatoagds** [comment]
@retrofox `20px` as padding doesn't follow our `8px` rule, updated the comment.

**Comment by retrofox** [comment]
> @retrofox `20px` as padding doesn't follow our `8px` rule, updated the comment.

We can use fractions:

```scss
    padding-top: calc( var(--spacing-base) * 2.5 ) );
```

**Comment by renatoagds** [comment]
@retrofox We don't use fractions based on `8 point` grid specifications. The only thing that we normally do is navigate between `8` to `4`.

Context:

https://uxplanet.org/everything-you-should-know-about-8-point-grid-system-in-ux-design-b69cb945b18d
https://medium.com/swlh/the-comprehensive-8pt-grid-guide-aa16ff402179

**Comment by keoshi** [comment]
> It's related too with the case that the padding applied is 20px at Figma, which doesn't follow our 8px rule

I'd be ok with going with 20px and breaking with our 8px multiples convention on this just because it gives the entire row more comfortable spacing.

> We're using from core, but it doesn't implement the margin-right, it's made by us, based on that, should I follow the 8px rule as normal? Which will move it to 32px (8 x 4) instead of 30px. Is that ok?

Noting that the icon is also contained in a frame like the thumbnail above, which gives it safe margins all around as visible in this image:

image

The logic for this quick actions menu was to align the “Edit video details” button to the second to last column, and let all icons flow after that separated by 16px.

image

**Comment by retrofox** [comment]
> @retrofox We don't use fractions based on `8 point` grid specifications. The only thing that we normally do is navigate between `8` to `4`.

Why we can't use 20px (8px * 2.5) to apply padding in the same way [we do with other elements](https://github.com/Automattic/jetpack/blob/trunk/projects/js-packages/components/components/button/style.module.scss#L29)?

Anyway, let's see what designer folks say.

**Comment by renatoagds** [comment]
> Why we can't use 20px (8px * 2.5) to apply padding in the same way [we do with other elements](https://github.com/Automattic/jetpack/blob/trunk/projects/js-packages/components/components/button/style.module.scss#L29)?

@retrofox As I said before and already put some context links, it goes out of 8 point grid system specification, it suggest you to navigate from `8px / 16px / 24px / 32px / ...etc` and just if needed go back for `4px`. So, only following the specification of the system that we decided to use :)

**Comment by renatoagds** [comment]
Based at our discussion here: p1661354188482839-slack-C03TA48NSUX

We'll follow with `16px` at top and bottom padding to make it consistent

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 1 above)
Comment (renatoagds) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: Done https://github.com/Automattic/jetpack/pull/25798/commits/838359bad832b622eaea67e2a4da7fcacd0bdd27

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 2 above)
Comment (renatoagds) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: Done https://github.com/Automattic/jetpack/pull/25798/commits/838359bad832b622eaea67e2a4da7fcacd0bdd27

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 3 above)
Comment (renatoagds) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: Done https://github.com/Automattic/jetpack/pull/25798/commits/838359bad832b622eaea67e2a4da7fcacd0bdd27

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 4 above)
Comment (renatoagds) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: Done https://github.com/Automattic/jetpack/pull/25798/commits/5a88689a0ba77c6cf2f833b44f43f0515dca5a99

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 5 above)
Comment (renatoagds) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx]: Done https://github.com/Automattic/jetpack/pull/25798/commits/5a88689a0ba77c6cf2f833b44f43f0515dca5a99

File: projects/packages/videopress/src/client/admin/components/video-row/index.tsx (hunk 8 above)
Comment (renatoagds) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx:182]: Good question. Don't know exactly, but we could figure out.

Comment (leogermani) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx:182]: Good catch. What we usually do is use the date format defined in the site settings

Comment (renatoagds) [reply projects/packages/videopress/src/client/admin/components/video-row/index.tsx:182]: @leogermani do you know how we do at client code? or @retrofox

**Review by retrofox — APPROVED** [review]
Thanks for addressing all feedback. LGMT. Nice work 👏 🚀

**Comment by renatoagds** [comment]
r251295-wpcom

**Pull request merged**